A Raft of Fixes To Private Dependencies

CodaFi opened the pull request:
- Turn the dependency verifier loose on the entire compiler pipeline, not just the typechecker
- Fixup the getter for the active dependency source. It was mistakenly set to read the source stack the old way when I refactored this in #31960 
- Make LookupInModuleRequest a dependency sink
- Commit a tricky regression test that verifies we always take a dependency on the underlying type of a typealias.
- Perform the unioning step during request replay.

Resolves rdar://64008262

CodaFi committed:
Verify The Entire Compilation Session

The old tests were just running the type checker because we used to only emit reference dependencies after Sema. Now that we emit them after the pipeline has run, let's upgrade these tests to capture these new references.

CodaFi committed:
Fix getActiveDependencySourceOrNull For Private Dependencies

Restore the behavior here of only looking at the topmost dependency source object that was lost when the referenced name tracker was removed.

Oops.

CodaFi committed:
Make LookupInModuleRequest a Dependency Sink

It wasn't one in the past because the referenced name tracker was written into by unqualified lookup, which would delegate to this request. Now that it's no longer doing that, we have to capture this edge here for private dependencies.

CodaFi committed:
Add Verifier Tests for A Tricky Typealias Miscompile

CodaFi commented:
@swift-ci smoke test

slavapestov commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -1692,8 +1692,7 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
     // Force requests that can emit diagnostics.
-    (void) TAD->getGenericSignature();
-    (void) TAD->getUnderlyingType();
+    (void) TAD->getInterfaceType();
```
This change should not be necessary

CodaFi committed:
Perform The Dependency Unioning Step During Replay

In order for private dependencies to be completely correct, it must perform the name lookup unioning step when a cached request is replayed - not just when lookups are first performed. In order to reduce the overhead of this union operation, it is not necessary to walk the entire active request stack, just walk to the nearest cached request in the stack and union into that. When it is popped, its replay step will itself union into the next cached request.

To see why, consider a request graph:

A* -> B -> C*
         |
          -> D*

where A, C, and D are cached.

If a caller were to force C and D, then force A independenty, today we would *only* replay the names looked up by C and D the first time A was evaluated. That is, subsequent evaluations of A do not replay the correct set of names. If we were to perform the union step during replay as well, requests that force A would also see C and D’s lookups.

Without this, callers that force requests like the DeclChecker have to be wary of the way they force the interface type request so other files see the right name sets.

rdar://64008262

CodaFi commented:
@swift-ci smoke test

CodaFi commented:
@swift-ci smoke test

CodaFi commented:
@swift-ci smoke test

CodaFi commented:
@swift-ci test Windows

CodaFi commented:
Just to see what happens here

@swift-ci test compiler performance

CodaFi commented:
@swift-ci test Windows

CodaFi commented:
@swift-ci test compiler performance

CodaFi commented:
@swift-ci test Windows

CodaFi commented:
(Oops, let's cancel the perf test. I meant to push a different stack of commits before that).

CodaFi commented:
⛵
